### PR TITLE
Fix bug when tapping ListTitle with CheckboxListTile  tristate enable

### LINKED ISSUE #63846: CheckboxListTile tristate: true not working
#58154 added `tristate` for `CheckboxListTile` but did not update the logic for what actually happens when the user taps on the list tile to call `onChanged`.

the documentation reads:

>  /// When a tri-state checkbox ([tristate] is true) is tapped, its [onChanged]
>  /// callback will be applied to true if the current value is false, to null if
>  /// value is true, and to false if value is null (i.e. it cycles through false
>  /// => true => null => false when tapped).

But the implementation does not support this:

https://github.com/flutter/flutter/blob/d1583fc39b8044d8e92fed63a01e0b523ce248b7/packages/flutter/lib/src/material/checkbox_list_tile.dart#L419

hence when trying to use it with tristate and `value: null` when tapping on the list tile produces:

```
flutter: 2020-08-15 12:14:41.735337 SHOUT main - Unhandled Flutter framework (gesture) error.
### Failed assertion: boolean expression must not be null
#0      CheckboxListTile.build.<anonymous closure> (package:flutter/src/material/checkbox_list_tile.dart:419:54)
#1      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:989:19)
#2      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:1098:38)
```

tapping on the Checkbox within the ListTile works as intended.

## Steps to Reproduce

create a `CheckboxListTile` with `tricheck: true`: Here an example: https://dartpad.dartlang.org/a50f547017bba61fc6c73f68aa100081

**Expected results:** 

true => null => false

**Actual results:** 

null error when `value` is null, otherwise:

true => false => tue => false

here the difference between tapping on the list tile vs tapping on the checkbox:

checkboxilsttile-tristate

### PR DESCRIPTION
**Pull request by xu-baolin**
## Description

fix bug when tapping `CheckboxListTile` ListTitle with tristate enable

## Related Issues

Fixes https://github.com/flutter/flutter/issues/63846

## Tests

I added the following tests:
`CheckboxListTile tristate test`

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by xu-baolin** [comment]
@HansMuller Please review, thanks :)

File: packages/flutter/test/material/checkbox_list_tile_test.dart [hunk 1]
```diff
@@ -171,18 +174,56 @@ void main() {
...
+    await tester.tap(find.byType(ListTile));
+    await tester.pumpAndSettle();
+    expect(_value, true);
+
+    await tester.tap(find.byType(ListTile));
+    await tester.pumpAndSettle();
+    expect(_value, false);
+
+    // Enable tristate
+    setState(()=> _tristate = true);
```
Comment (HansMuller) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:198]: It shouldn't be necessary to call setState() here, rebuilding the frame should be enough.

File: packages/flutter/test/material/checkbox_list_tile_test.dart [hunk 2]
```diff
@@ -171,18 +174,56 @@ void main() {
       ),
     );
 
-    expect(tester.widget<Checkbox>(find.byType(Checkbox)).value, null);
+    expect(tester.widget<Checkbox>(find.byType(Checkbox)).value, false);
+
+    // tap checkbox when disable tristate
```
Comment (HansMuller) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:179]: These comments are helpful. Wordsmithing: Tap the checkbox when tristate is disabled.

**Review by HansMuller — COMMENTED** [review]
I'm assuming that the call to `setState()` ("Enable tristate") was added to get the test to behave. Doesn't seem like it should be needed; if it is, lets understand exactly why.

File: packages/flutter/test/material/checkbox_list_tile_test.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/test/material/checkbox_list_tile_test.dart:198]: Yes, you are correct, it's redundant. I will fix it, thanks:)

Comment (xu-baolin) [reply packages/flutter/test/material/checkbox_list_tile_test.dart]: Done.

File: packages/flutter/test/material/checkbox_list_tile_test.dart (hunk 2 above)
Comment (xu-baolin) [reply packages/flutter/test/material/checkbox_list_tile_test.dart]: Done.

**Review by HansMuller — COMMENTED** [review]
This looks good, thanks for the contribution! Just asking for one small cosmetic change and then we can land the PR.

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**